File: week6_python/credit.py
# Asks the user for a credit card number and prints if it's valid and what type of card it is

def main():
    print("Number: ", end = "")
    number = input()

    while len(number) < 13 or len(number) > 16:
        print("The credit card number must have between 13 and 16 digits")
        print("Number: ", end = "")
        number = input()

    number = int(number)

    check_card(number)


def check_card(n):
    original_n = n

    # Calculate Luhn’s Algorithm
    sum_odd = 0
    sum_even = 0
    position = 0

    while n > 0:

        digit = n % 10
        if position % 2 == 0:
            sum_odd += digit
        else:
            product = digit * 2
            sum_even += (product // 10) + (product % 10) 
        n = n // 10
        position += 1

    
    # Checks to see if the card is valid
    if (sum_even + sum_odd) % 10 != 0:
        print("INVALID")
    else:
        while original_n >= 100:
            original_n = original_n // 10
    
        # Checks type of card
        if original_n == 34 or original_n == 37:
            print("AMEX")
        elif original_n // 10 == 4:
            print("VISA")
        elif original_n >= 51 and original_n <= 55:
            print("MASTERCARD")
        else:
            print("INVALID")





main()

File: week1_C/credit_cards/credit.c
// Asks for a credit card number and checks which type it is.

#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number;
    long original_number;

    // Asks the user for input
    do
    {
        number = get_long("Number: ");
        original_number = number; // Store the original number for later use
    }
    while (number <= 0);

    int sum_odd = 0; // Sum of digits in odd positions (from right)
    int sum_even = 0; // Sum of digits in even positions (from right)
    int position = 0; // To track the position of digits

    while (number > 0)
    {
        int digit = number % 10; // Get the last digit
        if (position % 2 == 0) // Odd position (considering the rightmost position as 0)
        {
            sum_odd += digit;
        }
        else // Even position
        {
            int product = digit * 2;
            sum_even += (product / 10) + (product % 10); // Add the digits of the product
        }
        number /= 10; // Remove the last digit
        position++;
    }

    // Check if the card is valid
    if ((sum_odd + sum_even) % 10 != 0)
    {
        printf("INVALID\n");
        return 1;
    }

    // Reduces the original number to the first two digits
    while (original_number >= 100)
    {
        original_number /= 10; 
    }
    
    // Determine card type based on original number
        if (original_number == 34 || original_number == 37) // AMEX (starts with 34 or 37)
    {
        printf("AMEX\n");
    }
    else if ((original_number / 10) == 4) // VISA (starts with 4)
    {
        printf("VISA\n");
    }
    else if (original_number >= 51 && original_number <= 55) // MASTERCARD (starts with numbers between 51 and 55)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n"); // If it doesn't match any known type
    }

    return 0;
}
